### PR TITLE
Feat/index on nullable collumns

### LINKED ISSUE #940: nullable columns in indexes
Implement indexes on columns with nullable values

- [x] change of disk format or storing indexing 
- [ ]  support null values indexes

Only the first one is needed for milestone 1.1.1

### PR DESCRIPTION
**Pull request by byo**
Closes #940

### CODE REVIEW

**Review by mmeloni — APPROVED** [review]

File: embedded/sql/engine.go [hunk 1]
```diff
@@ -924,7 +928,9 @@ func EncodeValue(val interface{}, colType SQLValueType, maxLen int) ([]byte, err
 		{
 			strVal, ok := val.(string)
 			if !ok {
-				return nil, ErrInvalidValue
+				return nil, fmt.Errorf(
+					"value is not a string: %w", ErrInvalidValue,
```
Comment (jeroiraz) [inline embedded/sql/engine.go:932]: what about "value %v is not a string: %w", val, ErrInvalidValue

File: embedded/sql/engine.go [hunk 2]
```diff
@@ -942,7 +948,9 @@ func EncodeValue(val interface{}, colType SQLValueType, maxLen int) ([]byte, err
 		{
 			intVal, ok := val.(int64)
 			if !ok {
-				return nil, ErrInvalidValue
+				return nil, fmt.Errorf(
+					"value is not an integer: %w", ErrInvalidValue,
```
Comment (jeroiraz) [inline embedded/sql/engine.go:952]: same as above

File: embedded/sql/engine.go [hunk 3]
```diff
@@ -957,7 +965,9 @@ func EncodeValue(val interface{}, colType SQLValueType, maxLen int) ([]byte, err
 		{
 			boolVal, ok := val.(bool)
 			if !ok {
-				return nil, ErrInvalidValue
+				return nil, fmt.Errorf(
+					"value is not a boolean: %w", ErrInvalidValue,
```
Comment (jeroiraz) [inline embedded/sql/engine.go:969]: same as above

File: embedded/sql/engine.go [hunk 4]
```diff
@@ -976,7 +986,9 @@ func EncodeValue(val interface{}, colType SQLValueType, maxLen int) ([]byte, err
 			if val != nil {
 				v, ok := val.([]byte)
 				if !ok {
-					return nil, ErrInvalidValue
+					return nil, fmt.Errorf(
+						"value is not a blob: %w", ErrInvalidValue,
```
Comment (jeroiraz) [inline embedded/sql/engine.go:990]: same as above

File: embedded/sql/engine.go [hunk 5]
```diff
@@ -1002,26 +1014,31 @@ func EncodeValue(val interface{}, colType SQLValueType, maxLen int) ([]byte, err
...
+	}
+
 	switch colType {
 	case VarcharType:
 		{
 			strVal, ok := val.(string)
 			if !ok {
-				return nil, ErrInvalidValue
+				return nil, fmt.Errorf(
+					"value is not a string: %w", ErrInvalidValue,
```
Comment (jeroiraz) [inline embedded/sql/engine.go:1040]: same as above

File: embedded/sql/engine.go [hunk 6]
```diff
@@ -1044,7 +1061,9 @@ func EncodeAsKey(val interface{}, colType SQLValueType, maxLen int) ([]byte, err
 
 			intVal, ok := val.(int64)
 			if !ok {
-				return nil, ErrInvalidValue
+				return nil, fmt.Errorf(
+					"value is not an integer: %w", ErrInvalidValue,
```
Comment (jeroiraz) [inline embedded/sql/engine.go:1065]: same as above

File: embedded/sql/engine.go [hunk 7]
```diff
@@ -1064,7 +1083,9 @@ func EncodeAsKey(val interface{}, colType SQLValueType, maxLen int) ([]byte, err
 
 			boolVal, ok := val.(bool)
 			if !ok {
-				return nil, ErrInvalidValue
+				return nil, fmt.Errorf(
+					"value is not a boolean: %w", ErrInvalidValue,
```
Comment (jeroiraz) [inline embedded/sql/engine.go:1087]: same as above

File: embedded/sql/engine.go [hunk 8]
```diff
@@ -1080,7 +1101,9 @@ func EncodeAsKey(val interface{}, colType SQLValueType, maxLen int) ([]byte, err
 		{
 			blobVal, ok := val.([]byte)
 			if !ok {
-				return nil, ErrInvalidValue
+				return nil, fmt.Errorf(
+					"value is not a blob: %w", ErrInvalidValue,
```
Comment (jeroiraz) [inline embedded/sql/engine.go:1105]: same as above

**Review by jeroiraz — DISMISSED** [review]
some unit testing covering nulls and joins would be useful, we may discover something...

**Comment by jeroiraz** [comment]
it may be simpler if we postpone merging this branch with fewer changes after transactions and header versioning are merged, also will simplify merging sessions branch...

**Comment by byo** [comment]
> some unit testing covering nulls and joins would be useful, we may discover something...

Added a simple test case, join was able to reference null value in foreign key.

**Pull request merged**